File: bindings/wasm/src/polyphony.ts
import { ErrorCode, SonareError } from './errors';
import { getSonareModule } from './module_state';
import type {
  NoteEditInput,
  NoteObject,
  PolyphonicAnalysisOptions,
  PolyphonicRenderOptions,
} from './public_types';
import type { WasmPolyphonicAnalysis } from './sonare.js';
import type { ValidateOptions } from './validation';
import { assertSampleRate, assertSamples } from './validation';

/** Canonical request form for {@link analyzePolyphonic}. */
export interface AnalyzePolyphonicRequest extends PolyphonicAnalysisOptions, ValidateOptions {
  samples: Float32Array;
  /**
   * Sample rate in Hz. Required: every duration in the chain — the ridge
   * minimum, the note minimum, the per-frame windows — is converted to samples
   * with this rate, so a wrong value silently analyses differently.
   */
  sampleRate: number;
}

/**
 * A polyphonic analysis, held as a handle so one note of a chord can be edited
 * and the result re-rendered without analysing the audio again.
 *
 * Created by {@link analyzePolyphonic}. **Release it with {@link destroy} as soon
 * as you are done with it**: the handle owns the source's complex spectrogram plus
 * the claimed bins of every note, which is the input over again plus the claims.
 * That is the price of re-rendering an edit for free, and it is why this is a
 * handle and not a result object — the measurement never crosses into JS.
 *
 * What crosses is what a host acts on: the notes ({@link notes}), each note's
 * pending edit ({@link setNoteEdit}), the per-frame voice count
 * ({@link polyphony}), and per note a pitch, a level and a salience curve. The
 * spectrogram, the masks and the per-bin weights do not, and no method reports a
 * per-bin figure.
 *
 * Using the analysis after it has been released throws `InvalidState` rather than
 * reaching a freed native object.
 */
export class PolyphonicAnalysis {
  private native: WasmPolyphonicAnalysis | null;

  /** Analyses the request's audio. {@link analyzePolyphonic} is the same call. */
  constructor(request: AnalyzePolyphonicRequest) {
    assertSamples('analyzePolyphonic', request.samples, request.validate !== false);
    assertSampleRate('analyzePolyphonic', request.sampleRate);
    const module = getSonareModule();
    this.native = module.createPolyphonicAnalysis(
      request.samples,
      request.sampleRate,
      request as unknown as Record<string, unknown>,
    );
  }

  private handle(): WasmPolyphonicAnalysis {
    if (this.native === null) {
      throw new SonareError(
        ErrorCode.InvalidState,
        'InvalidState',
        'PolyphonicAnalysis has been released',
      );
    }
    return this.native;
  }

  /** Number of notes, which is also the number of claim sets. */
  get noteCount(): number {
    return this.handle().noteCount;
  }

  /** Number of STFT frames the analysis ran over. */
  get frameCount(): number {
    return this.handle().frameCount;
  }

  /**
   * Every note, in the order their claim sets are held in — the same
   * {@link NoteObject} shape `extractNotes` returns, so a host that edits through
   * both doors sees one note.
   *
   * Each note carries its sample span, its frame span (in the analysis's own
   * framing), its median pitch, its steadiness, its per-frame `amplitude` and its
   * pending edit. The curves a note does not carry inline have their own accessors:
   * {@link noteF0}, {@link noteSalience}, and {@link noteEnvelope} for the points
   * last set through {@link setNoteEdit}. `amplitude` is {@link noteAmplitude}'s
   * curve, read once per note.
   */
  notes(): NoteObject[] {
    return this.handle().notes();
  }

  /**
   * Replaces one note's pending edit.
   *
   * The only thing a host writes. Everything else on a note is a measurement, and
   * the order is the pairing with the claim sets, so neither is settable.
   *
   * An omitted field is the identity, so `{}` restores the identity edit. The
   * envelope is `edit.amplitudeEnvelope`, which the handle copies, and its points
   * are per-frame linear gains over the note's span on top of `gainDb` — stretched
   * over whatever length the note renders at, so one entry is a constant gain and
   * the count need not match the note's frame count. Every value must be finite
   * and non-negative, which {@link render} is where it is checked, so one refusal
   * names one place.
   *
   * @param note - Index below {@link noteCount}
   * @throws {SonareError} `InvalidParameter` when `note` is out of range
   */
  setNoteEdit(note: number, edit: NoteEditInput): void {
    this.handle().setNoteEdit(note, edit);
  }

  /**
   * Voices estimated per frame, before tracking dropped anything — one entry per
   * frame from frame 0.
   *
   * What the estimation saw rather than what survived: a frame reported as three
   * voices with two notes spanning it is the difference between the two stages,
   * which is the figure a host deciding what to edit wants.
   */
  polyphony(): Int32Array {
    return this.handle().polyphony();
  }

  /**
   * One note's F0 in Hz, per frame over its own span.
   *
   * `frameEnd - frameStart` entries, so the value at index `i` belongs to frame
   * `frameStart + i`. This is the curve the monophonic door makes a caller pass
   * back in; here the handle already holds it, so a curve edit needs nothing from
   * the caller.
   *
   * @throws {SonareError} `InvalidParameter` when `note` is out of range
   */
  noteF0(note: number): Float32Array {
    return this.handle().noteF0(note);
  }

  /** One note's linear RMS, per frame over its own span. Indexed as {@link noteF0}. */
  noteAmplitude(note: number): Float32Array {
    return this.handle().noteAmplitude(note);
  }

  /**
   * One note's salience, per frame over its own span. Indexed as {@link noteF0},
   * and the one curve here that is not the note's own: it is the tracked ridge's,
   * so a frame of the note the ridge does not reach reads 0.
   *
   * Salience is what the estimation scored the candidate at, so it says how well
   * the material supported this note rather than how loud the note is —
   * {@link noteAmplitude} is the loud.
   */
  noteSalience(note: number): Float32Array {
    return this.handle().noteSalience(note);
  }

  /**
   * One note's amplitude envelope points, as last set — the same array
   * `notes()[note].edit.amplitudeEnvelope` carries.
   *
   * Indexed from 0 rather than over the note's span: an envelope is a set of gain
   * points stretched over whatever length the note renders at, not a per-frame
   * signal. The only one of the four curves that is not a measurement, and empty on
   * a note carrying no envelope.
   */
  noteEnvelope(note: number): Float32Array {
    return this.handle().noteEnvelope(note);
  }

  /**
   * Renders the analysis back to audio with whatever edits its notes carry, at the
   * source's length.
   *
   * Each note's claimed share is inverted, edited, and added to the residual — the
   * part of the input no note claimed. With every edit identity the result is the
   * analysis's own round trip, not the source bit for bit, the STFT round trip's
   * error being neither added to nor removed here.
   *
   * The render is additive per note with no cross-note term, so an unedited note's
   * contribution is identical between two renders. That is also the limit: a host
   * cannot tell from two renders whether a claim set divided the energy correctly.
   *
   * @throws {SonareError} `InvalidParameter` on an option or an edit field the
   *   renderer rejects — a non-positive stretch ratio, a negative or non-finite
   *   envelope point, or a vibrato or drift edit on a note carrying no usable
   *   pitch curve
   */
  render(options: PolyphonicRenderOptions = {}): Float32Array {
    return this.handle().render(options as Record<string, unknown>);
  }

  /**
   * Releases the underlying WASM object and everything it holds. A second call
   * throws `InvalidState` rather than freeing twice.
   */
  delete(): void {
    const native = this.handle();
    this.native = null;
    native.delete();
  }

  /** Alias for {@link delete}, provided for cross-binding (Node) compatibility. */
  destroy(): void {
    this.delete();
  }
}

/**
 * Analyses audio into editable notes and returns a handle to the analysis.
 *
 * One pass: one STFT, the multi-F0 extraction over it, a claim set per tracked
 * ridge, the apportionment of the bins two notes stand on, and the measured fields
 * of each note. Every note comes back with the identity edit, so rendering the
 * result unchanged reproduces the analysis's own round trip.
 *
 * An analysis finding no notes is not an error. Silence, or material the register
 * of the framing cannot resolve, tracks no ridge; rendering that is the residual
 * alone, which is the whole round trip.
 *
 * **This is for spans, not for whole songs.** The handle holds the source's
 * complex spectrogram plus every note's claimed bins — roughly 1,440 MiB for five
 * minutes of audio against a 2 GiB linear-memory cap — so analyse the passage you
 * are editing and {@link PolyphonicAnalysis.destroy} it when done.
 *
 * @throws {RangeError} on empty samples, a non-finite sample, or a `sampleRate`
 *   outside `[8000, 384000]`
 * @throws {SonareError} `InvalidParameter` on a config value the chain rejects
 *
 * @example
 * ```typescript
 * const analysis = analyzePolyphonic({ samples, sampleRate, maxPolyphony: 3 });
 * try {
 *   // Transpose the lowest note of the chord up a semitone.
 *   const notes = analysis.notes();
 *   const lowest = notes.reduce((a, b) => (a.medianHz <= b.medianHz ? a : b));
 *   analysis.setNoteEdit(notes.indexOf(lowest), { pitchShiftSemitones: 1 });
 *   const edited = analysis.render();
 * } finally {
 *   analysis.destroy();
 * }
 * ```
 */
export function analyzePolyphonic(request: AnalyzePolyphonicRequest): PolyphonicAnalysis {
  return new PolyphonicAnalysis(request);
}

File: src/wasm/bindings/editing/polyphony.cpp
/// @file polyphony.cpp
/// @brief Embind binding for polyphonic editing: one analysis held as a handle,
///        the notes it found, and a render back to audio.
///
/// The shape is sonare_c_polyphony.h's -- an opaque analysis and the operations
/// that header names -- and the call path is not: that C translation unit is not
/// linked into WASM, so every method reaches editing::polyphony directly and
/// resolves its own config and indices the way the C entry points do.
///
/// What stays inside the handle is the reason the handle exists. The analysis
/// holds the source's complex spectrogram and, per note, the complex weight of
/// every bin it claimed; neither crosses, and nothing here reports a per-bin
/// figure. What crosses is what a host acts on: the notes, each note's pending
/// edit, the per-frame voice count, and per note a pitch, a level and a salience
/// curve.
///
/// Nothing here catches -- a refusal is a SonareException, which the module wrapper
/// turns into a SonareError -- so there is no catch arm for emscripten's default
/// DISABLE_EXCEPTION_CATCHING to elide.

#ifdef __EMSCRIPTEN__

#include <algorithm>
#include <cmath>
#include <cstddef>
#include <string>
#include <vector>

#include "editing/note_model/note_renderer.h"
#include "editing/polyphony/polyphonic_edit.h"
#include "wasm/bindings/common/common.h"
#include "wasm/bindings/common/note_val.h"

namespace {

using sonare_wasm_editing::noteEditFromVal;
using sonare_wasm_editing::noteObjectsToVal;

// Mirrors sonare_c_polyphony.cpp's resolve_config: 0 takes the default on every
// field, and on the four that accept 0 as a value as well a negative selects it.
// Anything a stage range-checks is forwarded as given and refused there, so one
// bound is stated in one place.
editing::polyphony::PolyphonicEditConfig configFromVal(const val& config) {
  editing::polyphony::PolyphonicEditConfig out;
  // Absence and the default are this function's own; the conversion is the shared
  // reader's, and a value that cannot be a setting at all is refused here.
  const auto read = [&config](const char* key) {
    const float value = floatProperty(config, key, 0.0f);
    if (!std::isfinite(value)) {
      throw SonareException(ErrorCode::InvalidParameter,
                            std::string(key) + " must be a finite number");
    }
    return value;
  };
  const auto set_positive = [&read](const char* key, float& field) {
    const float value = read(key);
    if (value != 0.0f) field = value;
  };
  // The sentinel predicate is not what keeps a NaN out -- `value != 0.0f` is true of
  // one and `value < 0.0f` is not. Two layers refuse it: the read above, and
  // multi_f0.cpp, which re-checks finiteness on all four fields this writes.
  const auto set_floor = [&read](const char* key, float& field) {
    const float value = read(key);
    if (value < 0.0f) {
      field = 0.0f;
    } else if (value != 0.0f) {
      field = value;
    }
  };
  const auto set_count = [&config](const char* key, int& field) {
    const int value = intProperty(config, key, 0);
    if (value != 0) field = value;
  };

  auto& stft = out.extraction.stft;
  set_count("nFft", stft.n_fft);
  set_count("hopLength", stft.hop_length);
  set_count("winLength", stft.win_length);

  auto& spectrum = out.extraction.spectrum;
  set_positive("centRefHz", spectrum.ref_hz);
  set_positive("centsPerBin", spectrum.cents_per_bin);
  set_positive("centMaxHz", spectrum.max_hz);
  if (boolProperty(config, "tonalityOff", false)) spectrum.use_tonality = false;

  auto& salience = out.extraction.estimation.salience;
  set_count("salienceHarmonics", salience.n_harmonics);
  set_positive("f0MinHz", salience.f0_min_hz);
  set_positive("f0MaxHz", salience.f0_max_hz);
  set_positive("salienceAlphaHz", salience.alpha_hz);
  set_positive("salienceBetaHz", salience.beta_hz);
  // Zero is this field's default and also a stretch of zero, so it needs no
  // sentinel.
  salience.inharmonicity = read("salienceInharmonicity");

  auto& estimation = out.extraction.estimation;
  set_count("maxPolyphony", estimation.max_polyphony);
  set_floor("minFramePeakRatio", estimation.min_frame_peak_ratio);
  set_floor("minSeparationCents", estimation.min_separation_cents);
  set_positive("subtractionFactor", estimation.subtraction_factor);

  auto& ridges = out.extraction.ridges;
  set_positive("maxJumpCents", ridges.max_jump_cents);
  set_floor("minRidgePeakRatio", ridges.min_ridge_peak_ratio);
  set_floor("minRidgeDurationMs", ridges.min_duration_ms);

  set_count("maskHarmonics", out.masks.n_harmonics);
  set_positive("claimLobes", out.masks.claim_lobes);
  out.masks.inharmonicity = read("inharmonicity");

  out.estimate_inharmonicity = boolProperty(config, "estimateInharmonicity", false);
  set_count("inharmonicityMinPartials", out.inharmonicity.min_partials);
  set_positive("inharmonicityMaxResidualBins", out.inharmonicity.max_residual_bins);
  set_positive("inharmonicityMaxStretch", out.inharmonicity.max_inharmonicity);

  set_count("windowFrames", out.shared_bins.window_frames);
  set_positive("minPartialSeparation", out.shared_bins.min_partial_separation);
  set_positive("maxFitResidual", out.shared_bins.max_fit_residual);
  set_positive("maxWeightModulus", out.shared_bins.max_weight_modulus);
  // Zero already means "derive one" on this field.
  out.shared_bins.max_refine_hz = read("maxRefineHz");
  set_positive("f0ToleranceCents", out.shared_bins.f0_tolerance_cents);

  set_positive("segmentationThresholdCents", out.notes.segmenter.segmentation_threshold_cents);
  set_positive("minNoteMs", out.notes.segmenter.min_note_ms);
  set_positive("referenceHz", out.notes.segmenter.reference_hz);
  return out;
}

// The config is resolved before the samples are copied, the way the C entry point
// resolves it before run_offline.
editing::polyphony::PolyphonicAnalysis analyzeFromVal(val samples, int sample_rate, val config) {
  const editing::polyphony::PolyphonicEditConfig resolved = configFromVal(config);
  return editing::polyphony::analyze_polyphonic(loadValidatedAudio(samples, sample_rate), resolved);
}

}  // namespace

class PolyphonicAnalysisWasm {
 public:
  PolyphonicAnalysisWasm(val samples, int sample_rate, val config)
      : analysis_(analyzeFromVal(samples, sample_rate, config)) {}

  // A count crosses as a double: every other count on this surface does, and a
  // host sizes nothing from it -- the accessors below return whole arrays.
  double noteCount() const { return static_cast<double>(analysis_.notes.size()); }

  int frameCount() const { return analysis_.spectrum.n_frames(); }

  // The same note shape the by-value door returns, through the same converter: an
  // edit means one thing whichever door applies it, so a note has to read the same
  // way through both.
  val notes() const { return noteObjectsToVal(analysis_.notes); }

  void setNoteEdit(double note, val edit) {
    const std::size_t index = noteIndex(note);
    std::size_t cumulative_count = 0;
    // Whatever the render rejects about an edit it rejects on the way out, so a
    // field is not checked twice here and a refusal names one place.
    analysis_.notes[index].edit =
        noteEditFromVal(edit, "setNoteEdit edit", "setNoteEdit input", &cumulative_count);
  }

  val polyphony() const { return vectorToInt32Array(analysis_.track.polyphony); }

  val noteF0(double note) const {
    return vectorToFloat32Array(analysis_.notes[noteIndex(note)].f0_hz.values);
  }

  val noteAmplitude(double note) const {
    return vectorToFloat32Array(analysis_.notes[noteIndex(note)].amplitude.values);
  }

  // The one curve here that is not a measurement: the points a caller last handed
  // setNoteEdit, indexed from 0 rather than over the note's span. The same array a
  // note object carries inline; it is here because the C ABI's note struct reports
  // an envelope count and a count whose points cannot be fetched promises what it
  // cannot deliver.
  val noteEnvelope(double note) const {
    return vectorToFloat32Array(analysis_.notes[noteIndex(note)].edit.amplitude_envelope);
  }

  val noteSalience(double note) const {
    const std::size_t index = noteIndex(note);
    const editing::note_model::NoteObject& object = analysis_.notes[index];
    const std::vector<editing::polyphony::NoteMask>& masks = analysis_.masks.notes;
    if (index >= masks.size()) {
      throw SonareException(ErrorCode::InvalidState, "noteSalience: the note has no claim set");
    }
    const int ridge_index = masks[index].ridge_index;
    const std::vector<editing::polyphony::F0Ridge>& ridges = analysis_.track.ridges;
    if (ridge_index < 0 || static_cast<std::size_t>(ridge_index) >= ridges.size()) {
      throw SonareException(ErrorCode::InvalidState, "noteSalience: the claim set has no ridge");
    }

    // The ridge's curve, not the note's, so it is read through the ridge's own
    // start and a frame the ridge does not reach is 0 rather than the neighbour's
    // value.
    const editing::polyphony::F0Ridge& ridge = ridges[static_cast<std::size_t>(ridge_index)];
    std::vector<float> span(
        static_cast<std::size_t>(std::max(0, object.frame_end - object.frame_start)), 0.0f);
    for (std::size_t i = 0; i < span.size(); ++i) {
      const long long frame = static_cast<long long>(object.frame_start) +
                              static_cast<long long>(i) - static_cast<long long>(ridge.frame_start);
      if (frame < 0 || static_cast<std::size_t>(frame) >= ridge.salience.size()) continue;
      span[i] = ridge.salience[static_cast<std::size_t>(frame)];
    }
    return vectorToFloat32Array(span);
  }

  // Empty is the fit not having been asked for. Any other length would break the
  // per-note indexing this reports under, which an analysis built here cannot do.
  val noteInharmonicity() const {
    const std::vector<float>& fitted = analysis_.inharmonicity_fit;
    if (!fitted.empty() && fitted.size() != analysis_.notes.size()) {
      throw SonareException(ErrorCode::InvalidState,
                            "noteInharmonicity: the fit does not cover every note");
    }
    return vectorToFloat32Array(fitted);
  }

  val render(val options) const {
    // Resolved the way renderNotes resolves the same two fields, so one config
    // means one thing whichever door applies it.
    editing::note_model::NoteRenderConfig config;
    const float fade_ms = floatProperty(options, "fadeMs", 0.0f);
    const float vibrato_cutoff_hz = floatProperty(options, "vibratoCutoffHz", 0.0f);
    if (!std::isfinite(fade_ms) || fade_ms < 0.0f || !std::isfinite(vibrato_cutoff_hz) ||
        vibrato_cutoff_hz < 0.0f) {
      throw SonareException(ErrorCode::InvalidParameter,
                            "render: fadeMs and vibratoCutoffHz must be finite and non-negative");
    }
    if (fade_ms > 0.0f) config.fade_ms = fade_ms;
    if (vibrato_cutoff_hz > 0.0f) config.decomposition.vibrato_cutoff_hz = vibrato_cutoff_hz;

    const Audio rendered = editing::polyphony::render_polyphonic(analysis_, config);
    const std::vector<float> out(rendered.data(), rendered.data() + rendered.size());
    return vectorToFloat32Array(out);
  }

 private:
  // A note index arrives as a bare JS number, so it is checked before it indexes
  // anything: embind casts a JS number to size_t, which wraps a negative one into
  // a huge in-range-looking value.
  std::size_t noteIndex(double note) const {
    const std::size_t index = wasmIndexArg(note, "note index");
    if (index >= analysis_.notes.size()) {
      throw SonareException(ErrorCode::InvalidParameter, "note index is out of range");
    }
    return index;
  }

  editing::polyphony::PolyphonicAnalysis analysis_;
};

PolyphonicAnalysisWasm* createPolyphonicAnalysis(val samples, int sample_rate, val config) {
  return new PolyphonicAnalysisWasm(samples, sample_rate, config);
}

void registerPolyphonyBindings() {
  class_<PolyphonicAnalysisWasm>("PolyphonicAnalysis")
      .property("noteCount", &PolyphonicAnalysisWasm::noteCount)
      .property("frameCount", &PolyphonicAnalysisWasm::frameCount)
      .function("notes", &PolyphonicAnalysisWasm::notes)
      .function("setNoteEdit", &PolyphonicAnalysisWasm::setNoteEdit)
      .function("polyphony", &PolyphonicAnalysisWasm::polyphony)
      .function("noteF0", &PolyphonicAnalysisWasm::noteF0)
      .function("noteAmplitude", &PolyphonicAnalysisWasm::noteAmplitude)
      .function("noteSalience", &PolyphonicAnalysisWasm::noteSalience)
      .function("noteInharmonicity", &PolyphonicAnalysisWasm::noteInharmonicity)
      .function("noteEnvelope", &PolyphonicAnalysisWasm::noteEnvelope)
      .function("render", &PolyphonicAnalysisWasm::render);
  function("createPolyphonicAnalysis", &createPolyphonicAnalysis, allow_raw_pointers());
}

#endif  // __EMSCRIPTEN__
